File: AppLovin MAX Demo App - Java/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/applovin/nativeads/carouselui/support/AppLovinTouchToClickListener.java
package com.applovin.enterprise.apps.demoapp.ads.applovin.nativeads.carouselui.support;

import android.content.Context;
import android.view.MotionEvent;
import android.view.View;

public class AppLovinTouchToClickListener
        implements View.OnTouchListener
{
    /**
     * Max allowed duration for a "click", in milliseconds.
     */
    private static final int MAX_CLICK_DURATION = 1000;

    /**
     * Max allowed distance to move during a "click", in DP.
     */
    private static final int MAX_CLICK_DISTANCE = 10;

    private long  pressStartTime;
    private float pressedX;
    private float pressedY;

    private final Context              context;
    private final View.OnClickListener clickListener;

    public AppLovinTouchToClickListener(Context context, View.OnClickListener clickListener)
    {
        this.context = context;
        this.clickListener = clickListener;
    }

    @Override
    public boolean onTouch(View v, MotionEvent e)
    {
        switch ( e.getAction() )
        {
            case MotionEvent.ACTION_DOWN:
            {
                pressStartTime = System.currentTimeMillis();
                pressedX = e.getX();
                pressedY = e.getY();
                break;
            }
            case MotionEvent.ACTION_UP:
            {
                long pressDuration = System.currentTimeMillis() - pressStartTime;
                if ( pressDuration < MAX_CLICK_DURATION && distance( pressedX, pressedY, e.getX(), e.getY() ) < MAX_CLICK_DISTANCE )
                {
                    clickListener.onClick( v );
                }
            }
        }
        return true;
    }

    private float distance(float x1, float y1, float x2, float y2)
    {
        float dx = x1 - x2;
        float dy = y1 - y2;
        float distanceInPx = (float) Math.sqrt( dx * dx + dy * dy );
        return pxToDp( distanceInPx );
    }

    private float pxToDp(float px)
    {
        return px / context.getResources().getDisplayMetrics().density;
    }
}

File: AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/applovin/nativeads/carouselui/support/AppLovinTouchToClickListener.kt
package com.applovin.apps.kotlindemoapp.nativeads.carouselui.support

import android.content.Context
import android.view.MotionEvent
import android.view.View

class AppLovinTouchToClickListener(private val context: Context, private val clickListener: View.OnClickListener) : View.OnTouchListener
{
    /**
     * Max allowed duration for a "click", in milliseconds.
     */
    private val MAX_CLICK_DURATION = 1000

    /**
     * Max allowed distance to move during a "click", in DP.
     */
    private val MAX_CLICK_DISTANCE = 10

    private var pressStartTime: Long = 0
    private var pressedX: Float = 0.toFloat()
    private var pressedY: Float = 0.toFloat()

    override fun onTouch(v: View, e: MotionEvent): Boolean
    {
        when (e.action)
        {
            MotionEvent.ACTION_DOWN ->
            {
                pressStartTime = System.currentTimeMillis()
                pressedX = e.x
                pressedY = e.y
            }
            MotionEvent.ACTION_UP ->
            {
                val pressDuration = System.currentTimeMillis() - pressStartTime
                if (pressDuration < MAX_CLICK_DURATION && distance(pressedX, pressedY, e.x, e.y) < MAX_CLICK_DISTANCE)
                {
                    clickListener.onClick(v)
                }
            }
        }
        return true
    }

    private fun distance(x1: Float, y1: Float, x2: Float, y2: Float): Float
    {
        val dx = x1 - x2
        val dy = y1 - y2
        val distanceInPx = Math.sqrt((dx * dx + dy * dy).toDouble()).toFloat()
        return pxToDp(distanceInPx)
    }

    private fun pxToDp(px: Float): Float
    {
        return px / context.resources.displayMetrics.density
    }
}